File: src/main/java/org/kite9/diagram/batik/transform/TransformFactory.java
package org.kite9.diagram.batik.transform;

import org.kite9.diagram.batik.model.AbstractBatikDiagramElement;
import org.kite9.diagram.dom.transform.*;
import org.kite9.diagram.logging.Kite9ProcessingException;
import org.kite9.diagram.model.Leaf;
import org.kite9.diagram.model.style.ContentTransform;

public class TransformFactory {

	public static SVGTransformer initializeTransformer(AbstractBatikDiagramElement diagramElement, ContentTransform t, ContentTransform defaultTransform) {
		
		if (t == ContentTransform.DEFAULT) {
			t = defaultTransform;
		}
		
		// note we are using the fall-through approach here.
		switch (t) {
		case RESCALE:
		case CROP:
			if (diagramElement instanceof Leaf) {
				if (t==ContentTransform.RESCALE) {
					return new RescalingTransformer((Leaf) diagramElement);
				} else if (t==ContentTransform.CROP) {
					return new CroppingTransformer((Leaf) diagramElement);
				}
			}
			//$FALL-THROUGH$
		case POSITION:
			return new PositioningTransformer(diagramElement);
		case NONE:
			return new NoopTransformer();
		case DEFAULT:
		default:
			throw new Kite9ProcessingException("No transform defined for "+diagramElement);
		}
	}
}

File: k9layout/src/main/kotlin/org/kite9/diagram/dom/transform/TransformFactory.kt
package org.kite9.diagram.dom.transform

import org.kite9.diagram.logging.Kite9ProcessingException
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.Leaf
import org.kite9.diagram.model.style.ContentTransform

object TransformFactory {

	fun initializeTransformer(
        diagramElement: DiagramElement,
        t: ContentTransform,
        defaultTransform: ContentTransform
    ): SVGTransformer {
        var t = t
        if (t === ContentTransform.DEFAULT) {
            t = defaultTransform
        }
        return when (t) {
            ContentTransform.RESCALE, ContentTransform.CROP -> {
                if (diagramElement is Leaf) {
                    if (t === ContentTransform.RESCALE) {
                        return RescalingTransformer(diagramElement)
                    } else if (t === ContentTransform.CROP) {
                        return CroppingTransformer(diagramElement)
                    }
                }
                PositioningTransformer(diagramElement)
            }
            ContentTransform.POSITION -> PositioningTransformer(diagramElement)
            ContentTransform.NONE -> NoopTransformer()
            ContentTransform.DEFAULT -> throw Kite9ProcessingException("No transform defined for $diagramElement")
            else -> throw Kite9ProcessingException("No transform defined for $diagramElement")
        }
    }
}